sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,442 @@
+            if (_clustering_range_start + _current_read_size < _row_end) {
+                _cached_read = co_await data_read(std::min(_row_end - _current_read_size, _row_start), _row_end);
+            } else {
+                _cached_read = co_await data_read(_clustering_range_start, _row_end);
+            }
+            _current_read_size = std::min(max_read_size, _current_read_size * 2);
+        }
+        co_return make_buffer_input_stream(_cached_read.share(_cached_read.size() - len, len));
+    }
+    temporary_buffer<char> last_bytes(size_t len) {
+        auto tmp = _cached_read.share(_cached_read.size() - len, len);
+        _cached_read.trim(_cached_read.size() - len);
+        return tmp;
+    }
+
+    column_translation get_column_translation() {
+        return _sst->get_column_translation(_schema, _sst->get_serialization_header(), _sst->features());
+    }
+    void modify_cache(const row_body_skipping_context::tombstone_reversing_info& info) {
+        auto to_idx = [this] (uint64_t offset) {
+            return _cached_read.size() - (_row_end - offset);
+        };
+        _cached_read.get_write()[to_idx(info.kind_offset)] = 7 - (int)info.range_tombstone_kind;
+        if ((int)info.range_tombstone_kind < 2 || (int)info.range_tombstone_kind > 5) {
+            auto before = _cached_read.share(to_idx(info.before_offset), info.after_offset - info.before_offset);
+            auto after = _cached_read.share(to_idx(info.after_offset), _row_end - info.after_offset);
+            std::copy(after.begin(), after.end(), _cached_read.get_write() + to_idx(info.before_offset));
+            std::copy(before.begin(), before.end(), _cached_read.get_write() + to_idx(info.before_offset) + after.size());
+        }
+    }
+public:
+    partition_reversing_data_source_impl(const schema& s,
+            shared_sstable sst,
+            index_reader& ir,
+            uint64_t partition_start,
+            size_t partition_len,
+            reader_permit permit,
+            const io_priority_class& io_priority,
+            tracing::trace_state_ptr trace_state)
+        : _schema(s)
+        , _sst(std::move(sst))
+        , _ir(ir)
+        , _io_priority(io_priority)
+        , _permit(std::move(permit))
+        , _trace_state(std::move(trace_state))
+        , _partition_start(partition_start)
+        , _partition_end(partition_start + partition_len)
+        , _row_start(_partition_end)
+        , _row_end(_partition_end)
+    { }
+
+    partition_reversing_data_source_impl(partition_reversing_data_source_impl&&) noexcept = default;
+    partition_reversing_data_source_impl& operator=(partition_reversing_data_source_impl&&) noexcept = default;
+
+    virtual future<temporary_buffer<char>> get() override {
+        if (!_partition_header_context) {
+            _partition_header_context.emplace(data_stream(_partition_start, _partition_end), _partition_start, _partition_end - _partition_start, _permit);
```

Comment:
Perhaps use information from the promoted index to reduce the stream size, can be used to reduce overreads. Minor, we can do it in a later pass.